File: programming_methods/laboratories/5/1.py
"""
Задача №111623. Продавец

На плоскости заданы координаты n (4≤n≤20) разных вершин.

Найти кратчайший замкнутый маршрут, начинающийся и заканчивающийся в 1-й вершине и посещающий все остальные вершины по одному разу.
Разрешается (если так оказывается выгодно) «проезжать через вершину, не останавливаясь» (см. пример 1).

Длина маршрута считается как сумма длин составляющих его рёбер, длины отдельных рёбер считаются согласно обычной евклидовой метрике, как .

Входные данные

Первая строка содержит количество вершин n (4≤n≤20).
Каждая из следующих n строк содержит по два разделённых пробелом числа с плавающей точкой — x- и y-координаты соответствующей вершины.

Выходные данные

Первая строка должна содержать единственное число (с плавающей точкой) — найденную минимальную длину замкнутого тура.
Вторая строка должна содержать перестановку чисел 2, 3, ..., N — порядок, в котором надо посещать эти вершины.
Числа внутри второй строки должны разделяться одинарными пробелами.

РЕШЕНИЕ НЕ ПРОХОДИТ ПО СКОРОСТИ ИЗ-ЗА PYTHON.
"""
import math
from array import array
from collections import deque
from itertools import product
from typing import List, Tuple, Sequence, cast


def calculate_distance(
        u: int,
        v: int,
        coords: Sequence[Tuple[float, float]]
) -> float:
    """Вычисление расстояния между двумя вершинами."""
    return math.dist(coords[u], coords[v])


def initialize_dp_table(
        n: int,
        coords: Sequence[Tuple[float, float]]
) -> Sequence[array]:
    """Инициализация таблицы динамического программирования."""
    distance_table = [array('d', [float('inf')] * n) for _ in range(1 << n)]
    distance_table[1][0] = 0  # Начальная вершина посещена, длина пути 0

    for mask, u in product(range(1 << n), range(n)):
        if not (mask & (1 << u)):  # Если вершина u не посещена
            continue
        dp_mask_u = distance_table[mask][u]
        for v in filter(lambda vertex: not mask & (1 << vertex), range(n)):  # Генератор для вершин
            next_mask = mask | (1 << v)
            distance_table[next_mask][v] = min(distance_table[next_mask][v], dp_mask_u + calculate_distance(u, v, coords))

    return distance_table


def find_min_path_length(
        n: int,
        coords: Sequence[Tuple[float, float]],
        distance_table: Sequence[array]
) -> Tuple[float, int]:
    """Нахождение минимальной длины пути и конечной вершины."""
    last_mask = (1 << n) - 1
    return min((distance_table[last_mask][u] + calculate_distance(u, 0, coords), u) for u in range(1, n))


def reconstruct_path(
        n: int,
        coords: Sequence[Tuple[float, float]],
        distance_table: Sequence[array],
        end_vertex: int,
        last_mask: int
) -> Sequence[int]:
    """Восстановление пути из таблицы dp."""
    path = deque()

    mask, current_vertex = last_mask, end_vertex
    while current_vertex != 0:
        path.appendleft(current_vertex)
        current_distance = distance_table[mask][current_vertex]
        current_vertex = next(
            prev_vertex for prev_vertex in range(n)
            if (mask & (1 << prev_vertex)) and math.isclose(
                current_distance,
                distance_table[mask ^ (1 << current_vertex)][prev_vertex] +
                calculate_distance(prev_vertex, current_vertex, coords)
            )
        )
        mask ^= (1 << path[0])

    return path


def find_shortest_path(coords: Sequence[Tuple[float, float]]) -> Tuple[float, Sequence[int]]:
    """Основная логика для нахождения кратчайшего пути."""
    n = len(coords)

    dp = initialize_dp_table(n, coords)

    min_path_length, end_vertex = find_min_path_length(n, coords, dp)

    path = reconstruct_path(n, coords, dp, end_vertex, (1 << n) - 1)

    # Преобразование пути к нужному формату
    return min_path_length, [x + 1 for x in path]


def main() -> None:
    n = int(input())
    coords = cast(List[Tuple[int, int]], [tuple(map(float, input().split())) for _ in range(n)])
    min_path_length, min_path = find_shortest_path(coords)
    print(f"{min_path_length:.14e}")
    print(' '.join(map(str, min_path)))


if __name__ == "__main__":
    main()

File: programming_methods/laboratories/5/1.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <limits>
#include <deque>
#include <tuple>
#include <algorithm>

using namespace std;

class TravelingSalesmanGraph {
public:
    TravelingSalesmanGraph(const vector<pair<double, double>>& coordinates)
        : num_vertices(coordinates.size()), coordinates(coordinates) {
        distance_matrix.resize(1 << num_vertices, vector<double>(num_vertices, numeric_limits<double>::infinity()));
        distance_matrix[1][0] = 0.0; // Начальная вершина посещена, длина пути 0
        initialize_distance_matrix();
    }

    int get_num_vertices() const {
        return num_vertices;
    }

    pair<double, int> find_shortest_path_length() {
        int final_mask = (1 << num_vertices) - 1;
        double min_distance = numeric_limits<double>::infinity();
        int end_vertex = -1;
        for (int u = 1; u < num_vertices; ++u) {
            double distance = distance_matrix[final_mask][u] + calculate_distance(u, 0);
            if (distance < min_distance) {
                min_distance = distance;
                end_vertex = u;
            }
        }
        return {min_distance, end_vertex};
    }

    vector<int> reconstruct_path(int end_vertex, int final_mask) {
        deque<int> path;
        int mask = final_mask;
        int current_vertex = end_vertex;

        while (current_vertex != 0) {
            path.push_front(current_vertex);
            double current_distance = distance_matrix[mask][current_vertex];
            for (int prev_vertex = 0; prev_vertex < num_vertices; ++prev_vertex) {
                if ((mask & (1 << prev_vertex)) &&
                    abs(current_distance - (distance_matrix[mask ^ (1 << current_vertex)][prev_vertex] + calculate_distance(prev_vertex, current_vertex))) < 1e-9) {
                    current_vertex = prev_vertex;
                    mask ^= (1 << path.front());
                    break;
                }
            }
        }

        vector<int> result(path.begin(), path.end());
        return result;
    }

private:
    int num_vertices;
    vector<pair<double, double>> coordinates;
    vector<vector<double>> distance_matrix;

    void initialize_distance_matrix() {
        for (int mask = 0; mask < (1 << num_vertices); ++mask) {
            for (int u = 0; u < num_vertices; ++u) {
                if (!(mask & (1 << u))) continue;
                double current_distance = distance_matrix[mask][u];
                for (int v = 0; v < num_vertices; ++v) {
                    if (!(mask & (1 << v))) {
                        int next_mask = mask | (1 << v);
                        double new_distance = current_distance + calculate_distance(u, v);
                        distance_matrix[next_mask][v] = min(distance_matrix[next_mask][v], new_distance);
                    }
                }
            }
        }
    }

    double calculate_distance(int u, int v) {
        double dx = coordinates[u].first - coordinates[v].first;
        double dy = coordinates[u].second - coordinates[v].second;
        return sqrt(dx * dx + dy * dy);
    }
};

pair<double, vector<int>> find_optimal_tour(const vector<pair<double, double>>& coordinates) {
    TravelingSalesmanGraph graph(coordinates);
    auto [shortest_path_length, end_vertex] = graph.find_shortest_path_length();
    auto path = graph.reconstruct_path(end_vertex, (1 << graph.get_num_vertices()) - 1);
    for (int& vertex : path) {
        vertex += 1; // Преобразование индекса в 1-based
    }
    return {shortest_path_length, path};
}

int main() {
    int num_vertices;
    cin >> num_vertices;

    vector<pair<double, double>> coordinates(num_vertices);
    for (int i = 0; i < num_vertices; ++i) {
        cin >> coordinates[i].first >> coordinates[i].second;
    }

    auto [shortest_path_length, optimal_path] = find_optimal_tour(coordinates);

    cout.precision(14);
    cout << fixed << shortest_path_length << endl;
    for (size_t i = 0; i < optimal_path.size(); ++i) {
        if (i > 0) cout << " ";
        cout << optimal_path[i];
    }
    cout << endl;

    return 0;
}
